File: r1-spike/mobile/modules/loka-memory/ios/LokaMemoryModule.swift
import ExpoModulesCore

// A2 local module: OS-visible physical memory, evidence files the M1 pulls, and real process death.
public final class LokaMemoryModule: Module {
  public func definition() -> ModuleDefinition {
    Name("LokaMemory")
    Function("physicalMemory") { () -> [String: String] in
      ["api": "ProcessInfo.processInfo.physicalMemory", "bytes": String(ProcessInfo.processInfo.physicalMemory)]
    }
    // Documents/ in the app data container; `devicectl device copy from` reads it for a development-signed app.
    Function("writeFile") { (name: String, base64: String) throws in
      let dir = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
      try Data(base64Encoded: base64)!.write(to: dir.appendingPathComponent(name), options: .atomic)
    }
    // SIGABRT: the process dies and iOS writes a crash report whose native frames the dSYM symbolicates.
    Function("kill") { () -> Void in
      abort()
    }
  }
}

File: r1-spike/mobile/modules/loka-memory/android/src/main/java/expo/modules/lokamemory/LokaMemoryModule.kt
package expo.modules.lokamemory

import android.app.ActivityManager
import android.content.Context
import android.util.Base64
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.io.File

// A2 local module: OS-visible physical memory, evidence files the M1 pulls, and real process death.
class LokaMemoryModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("LokaMemory")
    Function("physicalMemory") {
      val am = appContext.reactContext!!.getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager
      val info = ActivityManager.MemoryInfo()
      am.getMemoryInfo(info)
      mapOf("api" to "ActivityManager.MemoryInfo.totalMem", "bytes" to info.totalMem.toString())
    }
    // App-specific external storage (/sdcard/Android/data/<package>/files): `adb pull` reads it
    // without run-as, which a non-debuggable release build does not allow.
    Function("writeFile") { name: String, base64: String ->
      val dir = appContext.reactContext!!.getExternalFilesDir(null)!!
      val tmp = File(dir, "$name.tmp")
      tmp.writeBytes(Base64.decode(base64, Base64.NO_WRAP))
      check(tmp.renameTo(File(dir, name))) { "rename failed: $name" }
    }
    // SIGKILL to this process: no handlers run, as when the OS kills the app.
    Function("kill") { android.os.Process.killProcess(android.os.Process.myPid()) }
  }
}
